arch/risc-v/src/bl602/bl602_pwm_lowerhalf.c

```diff
@@ -0,0 +1,473 @@
+                          uint32_t freq)
+{
+  uint16_t period = BL_PWM_CLK / freq;
+  uint16_t threshold1 = 0;
+  uint16_t threshold2 = 0;
+
+  pwm_channel_disable(chan);
+
+  /* Config pwm period and duty */
+
+  putreg32(threshold1, BL602_PWM_N_THRE1(chan));
+  putreg32(threshold2, BL602_PWM_N_THRE2(chan));
+  putreg32(period, BL602_PWM_N_PERIOD(chan));
+
+  pwm_channel_enable(chan);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: bl602_pwm_setup
+ *
+ * Description:
+ *   This method is called when the driver is opened.  The lower half driver
+ *   should configure and initialize the device so that it is ready for use.
+ *   It should not, however, output pulses until the start method is called.
+ *
+ ****************************************************************************/
+
+static int bl602_pwm_setup(FAR struct pwm_lowerhalf_s *dev)
+{
+  int i;
+  int ret = OK;
+  FAR struct bl602_pwm_s *priv = (FAR struct bl602_pwm_s *)dev;
+
+  UNUSED(i);
+
+#ifdef CONFIG_PWM_NCHANNELS
+  for (i = 0; i < CONFIG_PWM_NCHANNELS; i++)
+    {
+      bl602_configgpio(priv->chan_pin[i]);
+      pwm_init(i, BL_PWM_FREQ_DEFAULT);
+    }
+#else
+  bl602_configgpio(priv->chan_pin[0]);
+  pwm_init(0, BL_PWM_FREQ_DEFAULT);
+#endif
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: bl602_pwm_shutdown
+ *
+ * Description:
+ *   This method is called when the driver is closed.  The lower half driver
+ *   stop pulsed output, free any resources, disable the timer hardware, and
+ *   put the system into the lowest possible power usage state
+ *
+ ****************************************************************************/
+
+static int bl602_pwm_shutdown(FAR struct pwm_lowerhalf_s *dev)
+{
+  int i;
+  int ret = OK;
+  FAR struct bl602_pwm_s *priv = (FAR struct bl602_pwm_s *)dev;
+
+  UNUSED(i);
+
+#ifdef CONFIG_PWM_NCHANNELS
+  for (i = 0; i < CONFIG_PWM_NCHANNELS; i++)
+    {
+      pwm_channel_disable(priv->chan_pin[i]);
+      bl602_gpio_deinit(i);
+    }
+#else
+  pwm_channel_disable(priv->chan_pin[0]);
+  bl602_gpio_deinit(0);
+#endif
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: bl602_pwm_start
+ *
+ * Description:
+ *   (Re-)initialize the PWM and start the pulsed output
+ *
+ ****************************************************************************/
+
+static int bl602_pwm_start(FAR struct pwm_lowerhalf_s *dev,
+                           FAR const struct pwm_info_s *info)
+{
+  FAR struct bl602_pwm_s *priv = (FAR struct bl602_pwm_s *)dev;
+  int                     ret  = OK;
+  int                     i;
+
+  UNUSED(i);
+
+#ifdef CONFIG_PWM_NCHANNELS
+  for (i = 0; i < CONFIG_PWM_NCHANNELS; i++)
+    {
+      bl602_pwm_freq(priv, i, info->frequency);
+      bl602_pwm_duty(priv, i, info->channels[i].duty);
```

Comment:
Should the second argument have been `info->channels[i].channel`? Otherwise there's no way to change the order of channels.

Also PR #4227 added an option to break these loops when `channel == -1`, it looks like it was missed here.

Reply:
PR https://github.com/apache/incubator-nuttx/pull/5091 to add the early exit from PR #4227.

Reply:
and PR https://github.com/apache/incubator-nuttx/pull/5093 to use `chan` instead of `i`.